File: frontend/src/services/downloads/queue-get.ts
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";

import { log } from "@/lib/logger";

import type { APIResponse } from "@/types/api/api-response";
import type { DownloadQueueJob } from "@/types/database/download-queue";

export interface GetAllDownloadQueueItems_Response {
  jobs: DownloadQueueJob[];
}

export const GetAllDownloadQueueItems = async (): Promise<APIResponse<GetAllDownloadQueueItems_Response>> => {
  try {
    log("INFO", "API - Download Queue", "Fetch", "Fetching download queue entries");
    const response = await apiClient.get<APIResponse<GetAllDownloadQueueItems_Response>>(`/download/queue/item`);
    if (response.data.status === "error") {
      throw new Error(response.data.error?.message || "Unknown error fetching download queue entries");
    } else {
      log("INFO", "API - Download Queue", "Fetch", "Fetched download queue entries successfully", {
        jobs: response.data.data?.jobs,
      });
    }
    return response.data;
  } catch (error) {
    log("ERROR", "API - Download Queue", "Fetch", "Error fetching download queue entries", { error });
    return ReturnErrorMessage<GetAllDownloadQueueItems_Response>(error);
  }
};

File: backend/routing/download/queue_get.go
package routes_download

import (
	"aura/database"
	"aura/logging"
	"aura/utils/httpx"
	"net/http"
)

type GetAllDownloadQueueItems_Response struct {
	Jobs []database.DownloadQueueJob `json:"jobs"`
}

// GetAllDownloadQueueItems godoc
// @Summary      Download Queue - Get Items
// @Description  Retrieve the current items in the download queue, each with its status (pending, processing, success, warning, error). This endpoint allows clients to monitor the progress of queued download tasks and identify any issues that may have occurred during processing.
// @Tags         Download
// @Accept       json
// @Produce      json
// @Security     SessionCookie
// @Security     ApiKeyAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200  {object}  httpx.JSONResponse{data=GetAllDownloadQueueItems_Response}
// @Failure      500           {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/download/queue/item [get]
func GetAllDownloadQueueItems(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Download Queue - Get Items", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)
	var response GetAllDownloadQueueItems_Response

	jobs, Err := database.GetAllDownloadQueueJobs(ctx)
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	response.Jobs = jobs
	httpx.SendResponse(w, ld, response)
}
